pecos/core/xmc/inference.hpp

```diff
@@ -1127,51 +1128,17 @@ namespace pecos {
 
     // Prolongates the predictions of the previous layer to the select children of nodes.
     // The result is returned as a csr_t matrix.
-    csr_t prolongate_sparse_predictions(const csr_t& csr_pred, const csc_t& C, const csr_t& csr_codes) {
+    csr_t prolongate_sparse_predictions(const csr_t& csr_codes, const csc_t& C, const csr_t& selected_outputs_csr) {
         typedef typename csr_t::mem_index_type mem_index_type;
         typedef typename csr_t::index_type index_type;
         typedef typename csr_t::value_type value_type;
 
-        auto rows = csr_codes.rows;
-        auto cols = csr_codes.cols;
+        auto rows = selected_outputs_csr.rows;
+        auto cols = selected_outputs_csr.cols;
 
         // Compute the nnz's of each row
         mem_index_type* row_ptr = new mem_index_type[rows + 1];
-        row_ptr[0] = 0;
-#pragma omp parallel for schedule(dynamic,4)
-        for (index_type row = 0; row < rows; ++row) {
-            index_type row_nnz = 0;
-
-            mem_index_type csr_codes_row_start = csr_codes.row_ptr[row];
-            mem_index_type csr_codes_row_end = csr_codes.row_ptr[row + 1];
-
-            robin_hood::unordered_set<index_type> valid_cols;
-            valid_cols.reserve(csr_codes_row_end - csr_codes_row_start);
-            for (mem_index_type i = csr_codes_row_start; i < csr_codes_row_end; ++i) {
-                valid_cols.insert(csr_codes.col_idx[i]);
-            }
-
-            mem_index_type csr_pred_row_start = csr_pred.row_ptr[row];
-            mem_index_type csr_pred_row_end = csr_pred.row_ptr[row + 1];
-
-            for (mem_index_type i = csr_pred_row_start; i < csr_pred_row_end; ++i) {
-                mem_index_type C_col_start = C.col_ptr[csr_pred.col_idx[i]];
-                mem_index_type C_col_end = C.col_ptr[csr_pred.col_idx[i] + 1];
-
-                for (mem_index_type j = C_col_start; j < C_col_end; ++j) {
-                    if (valid_cols.count(C.row_idx[j])) {
-                        row_nnz++;
-                    }
-                }
-            }
-
-            row_ptr[row + 1] = row_nnz;
-        }
-
-        // Perform summation
-        for (index_type i = 0; i < rows; ++i) {
-            row_ptr[i + 1] += row_ptr[i];
-        }
+        std::memcpy(row_ptr, selected_outputs_csr.row_ptr, sizeof(mem_index_type) * (rows + 1));
```

Comment:
Since the value of `selected_outputs_csr` is not used and the result shares the same sparsity pattern. Would it be better to use `selected_outputs_csr` for storing the results instead of allocating new memory?

Reply:
Changes not added. Runtime tests have shown that using the sparsity pattern for returning results isn't significantly better and is often times slower. The only case where it is faster is for is_predict_only=True for multi thread real time inference.